core/touch_gesture.js

```diff
@@ -0,0 +1,287 @@
+/**
+ * @license
+ * PXT Blockly
+ *
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * https://github.com/Microsoft/pxt-blockly
+ *
+ * See LICENSE file for details.
+ */
+
+/**
+ * @fileoverview The class extends Blockly.Gesture to support pinch to zoom
+ * for both pointer and touch events.
+ * @author samelh@microsoft.com (Sam El-Husseini)
+ */
+'use strict';
+
+goog.provide('Blockly.TouchGesture');
+
+goog.require('Blockly.Gesture');
+
+goog.require('goog.asserts');
+goog.require('goog.math.Coordinate');
+
+
+/*
+ * Note: In this file "start" refers to touchstart, mousedown, and pointerstart
+ * events.  "End" refers to touchend, mouseup, and pointerend events.
+ */
+
+/**
+ * Class for one gesture.
+ * @param {!Event} e The event that kicked off this gesture.
+ * @param {!Blockly.WorkspaceSvg} creatorWorkspace The workspace that created
+ *     this gesture and has a reference to it.
+ * @constructor
+ */
+Blockly.TouchGesture = function(e, creatorWorkspace) {
+  Blockly.TouchGesture.superClass_.constructor.call(this, e,
+    creatorWorkspace);
+
+  /**
+   * Boolean for whether or not this gesture is a multi-touch gesture.
+   * @type {boolean}
+   * @private
+   */
+  this.isMultiTouch_ = false;
+
+  /**
+   * A map of cached points used for tracking multi-touch gestures.
+   * @type {Object<number|string, goog.math.Coordinate>}
+   * @private
+   */
+  this.cachedPoints_ = {};
+
+  /**
+   * A scale value tracking the previous gesture scale.
+   * @type {number}
+   * @private
+   */
+  this.previousScale_ = 0;
```

Comment:
This is the ratio between the starting distance between the touch points and the most recent distance between the touch points.  
Scales between 0 and 1 mean the most recent zoom was a zoom out.  
Scales above 1.0 mean the most recent zoom was a zoom in.